File: 백준/Silver/2630. 색종이 만들기/색종이 만들기.py
import sys

N = int(sys.stdin.readline())

arr = []

for i in range(N):
    nums = list(map(int, sys.stdin.readline().split(" ")))
    arr.append(nums)

white = 0
blue = 0

def recue(widthDot, heightDot, size):
    global white
    global blue
    sumNum = 0 # 합계확인
    if(size == 1):
        if(arr[heightDot][widthDot] == 1):
            blue += 1
        else:
            white += 1
    else:
        for i in range(size):
            sumNum += sum(arr[heightDot + i][widthDot : widthDot + size])
        if(sumNum == size * size):
            blue += 1
        elif(sumNum == 0):
            white += 1
        else:
            recue(widthDot, heightDot, size // 2)
            recue(widthDot + size // 2, heightDot, size // 2)
            recue(widthDot, heightDot + size // 2, size // 2)
            recue(widthDot + size // 2, heightDot + size // 2, size // 2)

recue(0, 0, N)

print(white)
print(blue)

File: 백준/Silver/2630. 색종이 만들기/색종이 만들기.java
import java.util.*;
import java.io.*;

public class Main{

    private static BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
    private static BufferedWriter bw = new BufferedWriter(new OutputStreamWriter(System.out));
    private static StringTokenizer st;
    private static int N;
    private static int nums[][];
    private static int blue, white = 0;
    

    private static void fun(int x, int y, int size) {
        int sum = 0;
        for(int i = y ; i < y + size ; i++) {
            for(int j = x ; j < x + size ; j++) {
                sum += nums[i][j];
            }
        }
        if(sum == size * size) {
            blue += 1;
        }
        else if(sum == 0) {
            white += 1;
        }
        else {
            fun(x, y, size / 2);
            fun(x + size / 2, y, size / 2);
            fun(x, y + size / 2, size / 2);
            fun(x + size / 2, y + size / 2, size / 2);
        }
    }



    public static void main(String[] args) throws IOException{

        N = Integer.parseInt(br.readLine());

        nums = new int[N][N];

        for(int i = 0 ; i < N ; i++) {
            st = new StringTokenizer(br.readLine(), " ");
            for(int j = 0 ; j < N ; j++) {
                nums[i][j] = Integer.parseInt(st.nextToken());
            }
        }
        
        fun(0, 0, N);

        bw.write(Integer.toString(white));
        bw.write("\n");
        bw.write(Integer.toString(blue));

        bw.flush();

        br.close();
        bw.close();
    }
}